ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/MaximumLagFilter.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.rest.server;
+
+import static java.util.Objects.requireNonNull;
+
+import com.google.common.collect.ImmutableMap;
+import io.confluent.ksql.execution.streams.RoutingFilter;
+import io.confluent.ksql.execution.streams.RoutingOptions;
+import io.confluent.ksql.rest.entity.LagInfoEntity;
+import io.confluent.ksql.rest.entity.QueryStateStoreId;
+import io.confluent.ksql.util.KsqlHostInfo;
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalLong;
+import java.util.function.Function;
+
+/**
+ * A RoutingFilter that filters hosts based upon changelog processing lag.
+ */
+public final class MaximumLagFilter implements RoutingFilter {
+
+  private final ImmutableMap<KsqlHostInfo, Optional<LagInfoEntity>> lagByHost;
+  private final RoutingOptions routingOptions;
+  private final OptionalLong maxEndOffset;
+
+  private MaximumLagFilter(
+      final RoutingOptions routingOptions,
+      final ImmutableMap<KsqlHostInfo, Optional<LagInfoEntity>> lagByHost,
+      final OptionalLong maxEndOffset
+  ) {
+    this.lagByHost = requireNonNull(lagByHost, "lagByHost");
+    this.routingOptions = requireNonNull(routingOptions, "requireNonNull");
+    this.maxEndOffset = requireNonNull(maxEndOffset, "maxEndOffset");
+  }
+
+  @Override
+  public boolean filter(final KsqlHostInfo hostInfo) {
+    final long allowedOffsetLag = routingOptions.getOffsetLagAllowed();
+    if (allowedOffsetLag >= 0) {
+      return lagByHost.getOrDefault(hostInfo, Optional.empty())
+          .map(hostLag -> {
+            // Compute the lag from the maximum end offset we've seen
+            final long endOffset = maxEndOffset.orElse(hostLag.getEndOffsetPosition());
+            final long offsetLag = Math.max(endOffset - hostLag.getCurrentOffsetPosition(), 0);
+            return offsetLag <= allowedOffsetLag;
+          })
+          // If we don't have lag info, we'll be conservative and include the host
+          .orElse(true);
```

Comment:
wouldn't the conservative thing be to exclude the host? if I asked for an offset lag of 0 and I get a host that just came online (so I don't have lag info yet) I'd be surprised by this!

Reply:
Vinoth and I were chatting about this a bit.  I think falling back on either value has its drawbacks.  If we defaulted to false, then we might start failing requests because we have no lags when we could have at least made an attempt to get some value they have.  On the other hand, it might be behind and could return overly stale data and we have no way of knowing.

I'm inclined to do false as you say.  I think the real solution is mitigating not having the data with these issues: #4471 #4472